### PR TITLE
Issue #2770747 by mglaman: Implement offer actions

### PR DESCRIPTION
**Pull request by mglaman**
Provides PromotionOffer plugins.

### CODE REVIEW

**Comment by mglaman** [comment]
#446 should tests!

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+use Drupal\Component\Plugin\ConfigurablePluginInterface;
+use Drupal\Component\Plugin\PluginInspectionInterface;
+use Drupal\Core\Executable\ExecutableInterface;
+use Drupal\Core\Plugin\ContextAwarePluginInterface;
+use Drupal\Core\Plugin\PluginFormInterface;
+
+/**
+ * Defines an interface for Offer plugins.
+ */
+interface PromotionOfferInterface extends ExecutableInterface, PluginFormInterface, ConfigurablePluginInterface, PluginInspectionInterface, ContextAwarePluginInterface {
```
Comment (mglaman) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php:16]: I'm not sure if it makes sense to extend core's `ExecutableInterface`. It just provides the execute method, but maybe we should provide our own method here that is `apply` or `run`, or something. And it can return a `bool` value so we know if it applied or not.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    return $this->getContextValue($this->appliesTo());
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyAdjustment(AdjustableInterface $item, Price $amount) {
+    $item->addAdjustment(new Adjustment([
+      'type' => 'promotion',
+      'label' => $this->pluginDefinition['label'],
```
Comment (mglaman) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php]: This is the plugin definition, we need to add a "Label" to the base configuration form. Or, if its on root promotion form make sure it sets it.

Actually, that's what should be done in https://www.drupal.org/node/2770731

**Comment by mglaman** [comment]
PHPCS and it broke the FunctionalJavascript tests, which need to be expanded anyways to set an offer.

File: modules/promotion/src/Annotation/PromotionOffer.php [hunk 3]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Drupal\commerce_promotion\Annotation;
+
+use Drupal\Component\Annotation\Plugin;
+
+/**
+ * Defines a Promotion Offer item annotation object.
```
Comment (bojanz) [inline modules/promotion/src/Annotation/PromotionOffer.php]: "Defines the promotion offer plugin annotation object."

File: modules/promotion/src/Annotation/PromotionOffer.php [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * Plugin namespace: Plugin\Commerce\PromotionOffer.
+ *
+ * @Annotation
+ */
+class PromotionOffer extends Plugin {
+
+
```
Comment (bojanz) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Double newline.

File: modules/promotion/src/Annotation/PromotionOffer.php [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   * @ingroup plugin_translatable
+   */
+  public $label;
+
+  /**
+   * The name of the module providing the type.
+   *
+   * @var string
+   */
+  public $module;
```
Comment (bojanz) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Each plugin has a 'provider' set, so this should be unneeded?

File: modules/promotion/src/Annotation/PromotionOffer.php [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+   * The array is keyed by context names.
+   *
+   * @var \Drupal\Core\Annotation\ContextDefinition[]
+   */
+  public $context = [];
+
+  /**
+   * The target this action applies to.
+   *
+   * For example, this should be 'order' or 'line_item'.
```
Comment (bojanz) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Feels weird for the possible targets to be up in the air. Maybe we make this $target_entity_type?

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    if ($this->configuration['amount'] != 0) {
+      /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+      $order = $this->getTargetEntity();
+
+      $calculated_amount = -($order->total_price->amount * $this->getAmount());
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php]: Might make sense to merge the Calculator class first, then rebase on top of it.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php (hunk 2 above)
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php]: Let's default it to t('Discount'), that's a safe option other systems (and discount 1.x) use, then and add a todo for #2770731.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+      'amount' => $amount,
+      'source_id' => $this->getPluginId(),
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php:51]: The source would be the ID of the promotion entity, not the plugin ID.

File: modules/promotion/src/PromotionOfferManager.php [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    // If the plugin did not specify a category, process that.
+    if (empty($definition['category'])) {
+      if ($target == PromotionOfferInterface::ORDER) {
+        $definition['category'] = $this->t('Order');
+      }
+      elseif ($target == PromotionOfferInterface::LINE_ITEM) {
+        $definition['category'] = $this->t('Product');
+      }
+      else {
+        $definition['category'] = $this->t('Other');
```
Comment (bojanz) [inline modules/promotion/src/PromotionOfferManager.php]: Is there such a thing as "Other"?

File: src/Plugin/Field/FieldType/PluginItemDeriver.php [hunk 10]
```diff
@@ -18,6 +18,7 @@ public function getDerivativeDefinitions($base_plugin_definition) {
     $supported = [
       'condition' => $this->t('Conditions'),
       'action' => $this->t('Action'),
+      'commerce_promotion_offer' => $this->t('Promotion Offer'),
```
Comment (bojanz) [inline src/Plugin/Field/FieldType/PluginItemDeriver.php]: We don't capitalize the second word in labels usually, no?

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/ProductPercentageOff.php [hunk 11]
```diff
@@ -0,0 +1,31 @@
+<?php
+
+namespace Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer;
+use Drupal\commerce_price\Price;
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/ProductPercentageOff.php]: Missing newline after namespace.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php [hunk 12]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer;
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php]: Missing newline after namespace.

File: modules/promotion/src/Annotation/PromotionOffer.php (hunk 6 above)
Comment (mglaman) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Because it made the annotation seem long and ugly. `targetEntityType` over `target`. But we can change it to being more verbose

File: modules/promotion/src/PromotionOfferManager.php (hunk 9 above)
Comment (mglaman) [inline modules/promotion/src/PromotionOfferManager.php]: Shipment? I had no idea what to do here because we can't throw an exception because what if people add promotions targeting other entities. Maybe I just need to load the entity definition and get the entity's label.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php (hunk 12 above)
Comment (mglaman) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php]: Fixed, this line just didn't change.

File: modules/promotion/src/Annotation/PromotionOffer.php (hunk 6 above)
Comment (mglaman) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Changing to targetEntityType, since we need an actual entity type plugin ID.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php (hunk 1 above)
Comment (mglaman) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php:16]: This is stupid, because that has to do with conditions and is a different PR to be made.

**Comment by mglaman** [comment]
It looks like core doesn't support fields that use shared storage and have serialized properties. So we'll need to move Offer to field storage item instead of base field.

sqlcontententitystorage_php_-_commerce2x_-____drupal_sites_commerce2x_

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php [hunk 13]
```diff
@@ -0,0 +1,28 @@
...
+ *   targetEntityType = "commerce_order",
+ * )
+ */
+class OrderPercentageOff extends PercentageOffBase {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function execute() {
+    if ($this->configuration['amount'] != 0) {
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/OrderPercentageOff.php]: Why would the settings form allow a 0 value to be accepted?

File: modules/promotion/src/Annotation/PromotionOffer.php [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+  public $context = [];
+
+  /**
+   * The target entity type this action applies to.
+   *
+   * For example, this should be 'commerce_order' or 'commerce_line_item'.
+   *
+   * @var string
+   */
+  public $targetEntityType;
```
Comment (bojanz) [inline modules/promotion/src/Annotation/PromotionOffer.php]: Annotation keys are snake_case, not camelCase. So it should be target_entity_type

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php [hunk 15]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer;
+
+use Drupal\Core\Form\FormStateInterface;
+
+/**
+ * Provides base class for percentage off offers.
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php]: Provides _the_ base class for percentage off offers.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  /**
+   * Gets the percentage amount, as a decimal, negated.
+   *
+   * @return string
+   *   The amount.
+   */
+  public function getAmount() {
+    return (string) (-$this->configuration['amount']);
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php]: Should the negation be done here, or in the code that creates the adjustment? Isn't it saner to say "This class provides a 19% off promotion" instead of "This class provides a -19% off promotion"?

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state) {
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php]: I think we usually do {} for empty methods?

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+   * @return \Drupal\Core\Entity\EntityInterface
+   *   The target entity.
+   */
+  public function getTargetEntity();
+
+  /**
+   * Applies the promotion offer's adjustment to an item.
+   *
+   * @param \Drupal\commerce_order\AdjustableInterface $item
+   *   The adjustable item.
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php]: This is odd. Why is it $item and not $entity?

File: modules/promotion/src/PromotionOfferManager.php [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\Core\Entity\EntityTypeManagerInterface
+   */
+  protected $entityTypeManager;
+
+  /**
+   * Constructor for OfferManager objects.
```
Comment (bojanz) [inline modules/promotion/src/PromotionOfferManager.php]: Constructs a new PromotionOfferManager object.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php (hunk 16 above)
Comment (mglaman) [reply modules/promotion/src/Plugin/Commerce/PromotionOffer/PercentageOffBase.php]: Well \Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\PromotionOfferBase::applyAdjustment already has a Price passed to it to pass to the Adjustment.

and in \Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\ProductPercentageOff::execute we invoke Multiple. I guess in the execute we can flip \Drupal\commerce_promotion\Plugin\Commerce\PromotionOffer\PercentageOffBase::getAmount.

Just felt like better DX.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php (hunk 18 above)
Comment (mglaman) [reply modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php]: Because we don't specify that AdjustableInterface extends any entity interface, and could therefore be anything

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php [hunk 20]
```diff
@@ -0,0 +1,103 @@
...
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applyAdjustment(AdjustableInterface $item, Price $amount) {
+    $item->addAdjustment(new Adjustment([
+      'type' => 'promotion',
+      // @todo Change to label from UI when added in #2770731.
+      'label' => t('Discount'),
+      'amount' => $amount->multiply('-1'),
```
Comment (mglaman) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferBase.php:50]: @bojanz this was my solution. so the promotion price is always inversed when creating the adjustment. 

So an offer will always return a price, but this ensures it gets deducted.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php (hunk 18 above)
Comment (bojanz) [reply modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php]: Perhaps that's our mistake.

Comment (mglaman) [reply modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php]: Should it extend EntityInterface, then? I mean in reality anyone could have an object that collects adjustments for some use case or another without it being a content entity.

File: modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+   *
+   * @return \Drupal\Core\Entity\EntityInterface
+   *   The target entity.
+   */
+  public function getTargetEntity();
+
+  /**
+   * Applies the promotion offer's adjustment to an item.
+   *
+   * @param \Drupal\commerce_order\EntityAdjustableInterface $item
```
Comment (bojanz) [inline modules/promotion/src/Plugin/Commerce/PromotionOffer/PromotionOfferInterface.php]: Now that it's EntityAdjustableInterface, we can rename $item to $entity :)

File: modules/order/src/EntityAdjustableInterface.php [hunk 22]
```diff
@@ -5,10 +5,10 @@
 /**
  * Defines an interface for objects that contain adjustments.
  *
- * @see \Drupal\commerce_order\Entity\OrderInterface
- * @see \Drupal\commerce_order\Entity\LineItemInterface
+ * @see \Drupal\commerce_order\Entity\OrderInterfaceEntity
+ * @see \Drupal\commerce_order\Entity\LineItemInterfaceEntity
  */
-interface AdjustableInterface {
+interface EntityAdjustableInterface {
```
Comment (bojanz) [inline modules/order/src/EntityAdjustableInterface.php]: extends EntityInterface?

File: modules/order/src/Entity/OrderInterface.php [hunk 23]
```diff
@@ -12,7 +12,7 @@
 /**
  * Defines the interface for orders.
  */
-interface OrderInterface extends AdjustableInterface, ContentEntityInterface, EntityChangedInterface, EntityOwnerInterface, EntityStoreInterface {
+interface OrderInterface extends EntityAdjustableInterface, ContentEntityInterface, EntityChangedInterface, EntityOwnerInterface, EntityStoreInterface {
```
Comment (bojanz) [inline modules/order/src/Entity/OrderInterface.php]: Now that the name has changed, let's move it after ContentEntityInterface

File: modules/order/src/Entity/LineItemInterface.php [hunk 24]
```diff
@@ -2,14 +2,14 @@
...
-use Drupal\commerce_order\AdjustableInterface;
+use Drupal\commerce_order\EntityAdjustableInterface;
 use Drupal\Core\Entity\EntityChangedInterface;
 use Drupal\Core\Entity\ContentEntityInterface;
 
 /**
  * Defines the interface for line items.
  */
-interface LineItemInterface extends AdjustableInterface, ContentEntityInterface, EntityChangedInterface {
+interface LineItemInterface extends EntityAdjustableInterface, ContentEntityInterface, EntityChangedInterface {
```
Comment (bojanz) [inline modules/order/src/Entity/LineItemInterface.php]: Now that the name has changed, let's move it after ContentEntityInterface

Comment (mglaman) [reply modules/order/src/Entity/LineItemInterface.php]: Diff didn't make this disappear, but moved.
